### PR TITLE
Update cache action and fix hashFiles bug

### LINKED ISSUE #37745: GitHub Action Build Caching Doesn't Work
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

The build cache uses the key `${{ runner.os }}-build-output-${{ hashFiles('node_modules/.cache/turbo/*-meta.json') }}`. For some reason, however, the cache is not being updated from subsequent runs. It looks like `Linux-build-output-` is seen as the key and triggering `Cache hit occurred on the primary key Linux-build-output-, not saving cache.` on save.

This means that the cache is _never updated_. It gets set in `trunk` as `Linux-build-output-` a long time ago and never gets set again. The branch pulls this in, uses the same `Linux-build-output-` key, can't update the cache, and so it isn't caching anything.

--

In addition to that, we should take a look at the caching behavior and figure out which cache it pulls from the fallback `restore-key` of `Linux-build-output-`. Since the hash is made from all of the package hashes, if one package changes, the entire cache is invalidated. This is fine as long as the restore key is pull the _latest_ cache. 

- `trunk` has a cache with some `Linux-build-output-hash` key.
- Create a PR from `trunk`, it will pull the cache from `trunk` just fine.
  - If it has the latest cache from `trunk`, this will work fine. It will only have cache misses for the packages that have changed.
  - If it has a random cache from `trunk`, we can't be sure how aggressive the cache will miss.
- PR builds and updates the cache for the branch with `Linux-build-output-new-hash`.
  - If you make another change, when it pulls from `Linux-build-output-` as the fallback, will it restore the last cache from the branch?

Does it pull from the latest cache in a given branch when restoring from `Linux-build-output-`?

### Expected behavior

The build cache for branches should be populated with the most-relevant cache available.

### Actual behavior

The build cache does not work at all.

### Steps to reproduce

1. Look at any PR's checks, note that the "Setup WooCommerce Monorepo" action never has any cache hits.

### WordPress Environment

N/A

### Isolating the problem

- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [ ] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by samueljseay**
### Changes proposed in this Pull Request:

This adds 2 simple changes to fix https://github.com/woocommerce/woocommerce/issues/37745

1. This updates the cache action used in the monorepo setup GH action to use the latest released version. There have been many bug fixes and updates since this action was last updated. This is not needed to fix the problem, but I wanted to update the action while I was working in this area.

2. Fixes a bug where the combination of `path` and matcher passed to `hashFiles` resulted in the hash being empty and there was no cache restoration happening. `hashFiles` will return an empty string whenever the matcher matches no files.

To confirm the fix, I pushed a change that should not bust the cache after the first commit.

Now you can see the output being correct and that the cache is being restored from a result of hashing the turbo meta.json files.

e.g. I saw this output for 2 runs of the CI:

First run of CI that generated the now correct cache key:

`key: Linux-build-output-07f6725759c35c31ec851eeede02b20ee638aa198b7e1716bbd2b3d2d74265c9`

Second run, restored the cache due to their being no changes:

`Cache restored from key: Linux-build-output-07f6725759c35c31ec851eeede02b20ee638aa198b7e1716bbd2b3d2d74265c9`

### How to test the changes in this Pull Request:

1. It is sufficient to see that all the Github CI checks pass and to see that the "WooCommerce MonoRepo Setup" steps used in many of the checks is successfully restoring the cache with output like: `Cache restored from key: Linux-build-output-07f6725759c35c31ec851eeede02b20ee638aa198b7e1716bbd2b3d2d74265c9`

### CODE REVIEW

**Comment by ObliviousHarmony** [comment]
Thanks @samueljseay,

While I _do_ see that the cache key is being populated now, it doesn't actually look like it's doing what it says on the tin. I've re-ran the unit test checks a few times and all of the package builds are cache misses each time.

**Comment by samueljseay** [comment]
@ObliviousHarmony oh wow, you want to have your cake and eat it too? 😛 I'll investigate why the build is not cached. thanks.

**Comment by samueljseay** [comment]
@ObliviousHarmony so I think I'm closer to understanding the problems now, I've come up with a simplified solution that does seem to produce cache hits on turbo build.

Some problems and thoughts:

1. Because the hashFiles was an empty string, we had one very useless cache being pulled up a lot by the action trying to restore the cache. The cache was very old and obviously contained no cache hits for turbo. Also I think my fix that did seem to generate a hashFiles string was still useless, it did not contain a relevant set of turbo cache so it could have been a red herring.

2. The old way we were trying to cache was never going to work well because hashFiles is calculated with files that don't exist yet. I can't think of a way that hashFiles could ever work properly with the way we were doing it before. The `*-meta.json` files are generated as a result of the build, so even if we correctly called hashFiles when they existed there would be no way restore the cache again before running build in future. It'd be cool if you could run turbo and just get a single hash for the whole workspace that represents whether anything changed without actually running anything, but it doesn't support that in the CLI right now. I imagine thats the output I would use to generate the key. If it was generated fast enough it would probably make Github actions local caching with turbo much simpler.

3. So what I've done instead is make hashFiles an approximation of the build inputs in our turbo repo config. Perhaps it could be improved more than what I've provided here, but this hashFiles call will generate the cache in a fashion that is "close" to what turbo would bust the cache for. this hashFiles calculation works because all these files exist when we try hashFiles before the build happens. If the hashFiles calculation determines nothing relevant has changed it will restore the cache and if Turbo finds the hashes it wants in that restored cache then it will get cache hits for the relevant packages. This did result in a much faster subsequent build in testing.

4. Note that 3 is certainly not perfect, there will be situations where it will restore a cache that will result in misses. Can I improve the inputs to `hashFiles` in your opinion?

5. Restore keys should be kept I guess? At least any hashes that happen to be in the closest cache we can find will speed up the building of those packages, but it may also hide problems with caching like old stale caches being used.

**Comment by ObliviousHarmony** [comment]
Thanks for such a deep investigation @samueljseay! You make a _great_ point about the non-existent build meta files. Why did it appear like the cache ever worked at all? Is the new cache action a different major version?

I don't actually think we should tie the cache key to any file-related state. When we do that, if _any_ file changes, we lose the entire build cache for that run. Even if only a single package changed we will rebuild them all. With respect to our build caching I think a little bit of a cache is better than no cache at all. With that in mind, I've got a few questions and thoughts:

- Can we replace the cache if the key is already present, or, do we need to register a new cache key?
- Should we use the branch name in the cache key instead of file hashes? This would mean subsequent pushes only rebuild things changed in the most recent commit(s).
  - If we can't replace the cache for the branch, maybe we can use the branch as a restore key and use the commit hash as the most-specific part of the key?

Because of the way cache restoration works in GitHub Actions, this can do some nice things for us:

- `${{ runner.os }}-build-output-` as a restore key will, generally, only be useful when initially creating the PR. Since there's no branch-specific cache at this point, it will pull from `trunk` and get us a cache populated that will only build things changed by the PR. When the build completes it'll store a cache to use in the branch from then on.
- `${{ runner.os }}-build-output-<BRANCH HERE>-` will be the branch-specific restore key. If we can replace the cache instead of adding to it, this is where we can stop and this will just be the cache key (without the trailing hyphen).
- `${{ runner.os }}-build-output-<BRANCH HERE>-<COMMIT HASH>` would be the cache key. This will change with every HEAD commit once they push the branch. This is only necessary though if we can't replace the existing cache key for the branch.

This would probably get us the highest number of cache hits out of Turbo. What do you think?

**Comment by samueljseay** [comment]
@ObliviousHarmony your suggestion is actually very close to the one used here: https://github.com/vercel/turbo/pull/2761/files 

They don't limit it to the branch, but rather just the commit hash for the current PR with a restore key that all branches can use. Isn't that better? Because we should in theory have a lot of situations where the first commit of a PR can get a partially cached first build? 

> Why did it appear like the cache ever worked at all? Is the new cache action a different major version?

This is a major version change yes with many fixes and behaviour changes, I can't frankly answer why it worked or appeared to in the past without doing a lot of investigation that feels like not a good use of time.


> I don't actually think we should tie the cache key to any file-related state. When we do that, if any file changes,

Your statement here confuses me a bit. I don't think this is true for *any* file? Only if one of the file patterns passed to hashFiles changes will we get a fresh cache generated and we still have a restore key (restore key is basically a prefix for searching for the closest matching last cache) which will search existing caches that are shared so you'll often likely get some cache hits anyway from other PRs?

**Comment by ObliviousHarmony** [comment]
@samueljseay that does look similar.

> They don't limit it to the branch, but rather just the commit hash for the current PR with a restore key that all branches can use. Isn't that better?

I was thinking a key that looks like`Linux-build-output-dev/update-cache-action-e7407e6fd3ebab747c9bf2955acd7075675a2025` with restore keys `Linux-build-output-dev/update-cache-action-` and `Linux-build-output-`. Consider the following:

- PR opened changing `@woocommerce/components` and `woocommerce/client/admin`.
- Cache hit from `trunk` containing matching Turbo cache entries for everything but those two packages.
- PR's checks run and only rebuild `@woocommerce/components` and `woocommerce/client/admin`.
- Author changes `woocommerce/client/admin`.
- PR's checks run again.

If we're _only_ using the commit hash, every time the checks run they will rebuild anything that has changed from the first commit. Even if the change between two commits in the same PR would build less (or in some cases, such as documentation, nothing at all!), it will build anything changed in that PR. It's true that this difference may be minuscule, but, if we're storing a cache for every `git push` anyway, we may as well have it contain the latest Turbo outputs.

> Only if one of the file patterns passed to hashFiles changes will we get a fresh cache generated

Keep in mind that there's an important distinction here between the granularity of GitHub Action's cache and Turbo's cache. If we use the input files in `hashFiles()`, if any file changes, it will invalidate the _entire_ cache. Rather than taking advantage of a partial rebuild from Turbo we're going to rebuild the entire dependency graph.

> will search existing caches that are shared so you'll often likely get some cache hits anyway from other PRs?

You won't ever get a cache hit from another PR. For security purposes you are only able to access the cache of your base branch (and any of its base branches). For example: in the configuration `PR2 -> PR1 -> trunk`, `PR2` can only access caches from itself, `PR1`, and `trunk`.

**Comment by samueljseay** [comment]
> You won't ever get a cache hit from another PR. For security purposes you are only able to access the cache of your base branch (and any of its base branches). For example: in the configuration PR2 -> PR1 -> trunk, PR2 can only access caches from itself, PR1, and trunk.

ok I see, I didn't understand this. I could not determine how cache isolation worked but that makes sense.

> I was thinking a key that looks likeLinux-build-output-dev/update-cache-action-e7407e6fd3ebab747c9bf2955acd7075675a2025 with restore keys Linux-build-output-dev/update-cache-action- and Linux-build-output-. 

Yeah ok, I think that makes sense to me. Thanks

**Review by ObliviousHarmony — COMMENTED** [review]
I took a peek at [GitHub's environment variables](https://docs.github.com/en/actions/learn-github-actions/variables) and it documents a `GITHUB_REF_NAME` variable that seems like it might be fit for purpose. I noticed that you tried this out in c1cc6afa3ebc95c8b664d5d241a832a84bb839bf, but, I'm curious why you went with the composite action instead?

**Comment by samueljseay** [comment]
@ObliviousHarmony I found that variable was not matching what the documentation said it should. I would get values like `refs/pull/n/merge`. I had a quick look just now and actually this issue looks like the problem i saw based on event type: https://github.com/github/docs/issues/15319

TLDR; the documentation is wrong.

After much searching the simplest thing comes up a lot, basically set the variable based on event type. And I made this action so we can reuse it because it seems useful

File: .github/actions/set_branch_name/action.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+description: Sets the BRANCH_NAME env variable based on the Github Event
+permissions: {}
+
+runs:
+    using: composite
+    steps:
+        - name: Get branch name (merge)
+          if: github.event_name != 'pull_request'
+          shell: bash
+          run: echo "BRANCH_NAME=$(echo ${GITHUB_REF#refs/heads/} | tr / -)" >> $GITHUB_ENV
```
Comment (ObliviousHarmony) [inline .github/actions/set_branch_name/action.yml]: In the non-branch case, `GITHUB_REF_NAME` should work, right?

File: .github/actions/set_branch_name/action.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+name: Set BRANCH_NAME environment variable
```
Comment (ObliviousHarmony) [inline .github/actions/set_branch_name/action.yml]: nit: Do you think `BRANCH_NAME` might cause problems eventually since it's an environment variable made accessible to anything in the environment?

**Review by ObliviousHarmony — DISMISSED** [review]
> TLDR; the documentation is wrong.

Great work finding a solution that works :)

> I made this action so we can reuse it because it seems useful

Do you think that might be premature?

---

The cache seems to populate correctly now, thanks @samueljseay. These comments are just nits, so, feel free to merge if you want. Works great!

File: .github/actions/set_branch_name/action.yml (hunk 2 above)
Comment (samueljseay) [reply .github/actions/set_branch_name/action.yml]: @ObliviousHarmony 

> Do you think that might be premature?

Hmm yes maybe.

I have this thing bothering me, that we tend to end up having the same things reimplemented multiple times in disparate actions and I want to encourage people to reuse existing ones, but that said I don't think me introducing this will make anyone reuse it, they'll need to know it exists and as you imply, we don't know if anyone needs it yet anyway.

> nit: Do you think BRANCH_NAME might cause problems eventually since it's an environment variable made accessible to anything in the environment?

I think you're right, I'll bring this into the workflow instead and we won't need an env variable anyway 😃

File: .github/actions/set_branch_name/action.yml (hunk 1 above)
Comment (samueljseay) [reply .github/actions/set_branch_name/action.yml]: ah yes true. yes that would simplify this.

**Review by ObliviousHarmony — DISMISSED** [review]
Looks great @samueljseay, thanks for fixing our cache! I'm sure people will really appreciate the faster build times.

Do you think a follow-up to this pull request that populates the cache before the rest of the actions run might make sense? Technically, while there's a cache, every workflow is going to get a partial cache miss and do a rebuild.

**Comment by samueljseay** [comment]
@ObliviousHarmony sadly turborepo caching is not working now (all of a sudden!). As far as I can tell the cache is being created and restored correctly but turborepo is cache missing. I'll have to dig more into this today 😭

**Comment by samueljseay** [comment]
@ObliviousHarmony I think this is fixed now :D I see cache hits happening properly.

The change I made to fix it is to bring the cache dir outside of node_modules. I don't actually know the *exact* reason why having this change fixes it, because in my mind the cache should be restorable to and from node_modules. Especially because we do indeed create/restore the cache *after* node module install.

All of that said, many others have faced this issue. see: https://github.com/vercel/turbo/issues/451

I have theories about it, like maybe the cache restoration in pnpm is causing issues with.. file permissions creating the directory? that kind of thing? ha I'm not sure sadly. There is a secondary problem that the cache action doesn't seem to give any helpful information about what's happening (like if it can't find matching files to cache or similar)

Anyway after moving the turbo cache out of node_modules using `--cache-dir` I noticed that caches immediately started functioning properly.

**Comment by samueljseay** [comment]
@ObliviousHarmony 

> Do you think a follow-up to this pull request that populates the cache before the rest of the actions run might make sense? Technically, while there's a cache, every workflow is going to get a partial cache miss and do a rebuild.

Yes its a good idea! I'll create a task to follow up with that. Thank you for the suggestion.

**Review by ObliviousHarmony — DISMISSED** [review]
The caches seem to work, so, no problems here :smile: Great work @samueljseay!

**Review by ObliviousHarmony — APPROVED** [review]
